src/eva/vision/data/datasets/segmentation/total_segmentator.py

```diff
@@ -148,23 +160,54 @@ def __len__(self) -> int:
     def load_image(self, index: int) -> tv_tensors.Image:
         sample_index, slice_index = self._indices[index]
         image_path = self._get_image_path(sample_index)
-        image_array = io.read_nifti_slice(image_path, slice_index)
+        image_array = io.read_nifti(image_path, slice_index)
         if self._as_uint8:
             image_array = convert.to_8bit(image_array)
         image_rgb_array = image_array.repeat(3, axis=2)
         return tv_tensors.Image(image_rgb_array.transpose(2, 0, 1))
 
     @override
     def load_mask(self, index: int) -> tv_tensors.Mask:
+        if self._optimize_mask_loading:
+            return self._load_semantic_label_mask(index)
+        return self._load_mask(index)
+
+    def _load_mask(self, index: int) -> tv_tensors.Mask:
+        """Loads and builds the segmentation mask from NifTi files."""
         sample_index, slice_index = self._indices[index]
         masks_dir = self._get_masks_dir(sample_index)
-        mask_paths = (os.path.join(masks_dir, label + ".nii.gz") for label in self.classes)
-        binary_masks = [io.read_nifti_slice(path, slice_index) for path in mask_paths]
-        one_hot_encoded = np.concatenate(binary_masks, axis=2)
-        background_mask = one_hot_encoded.sum(axis=2, keepdims=True) == 0
-        one_hot_encoded_with_bg = np.concatenate([background_mask, one_hot_encoded], axis=2)
-        segmentation_label = np.argmax(one_hot_encoded_with_bg, axis=2)
-        return tv_tensors.Mask(segmentation_label)
+        mask_paths = [os.path.join(masks_dir, label + ".nii.gz") for label in self.classes]
+        binary_masks = [io.read_nifti(path, slice_index) for path in mask_paths]
+        background_mask = np.zeros_like(binary_masks[0])
+        semantic_labels = np.argmax([background_mask] + binary_masks, axis=0)
+        return tv_tensors.Mask(semantic_labels)
+
+    def _load_semantic_label_mask(self, index: int) -> tv_tensors.Mask:
+        """Loads the segmentation mask from a semantic label NifTi file."""
+        sample_index, slice_index = self._indices[index]
+        masks_dir = self._get_masks_dir(sample_index)
+        filename = os.path.join(masks_dir, "semantic_labels", "masks.nii.gz")
+        semantic_labels = io.read_nifti(filename, slice_index)
+        return tv_tensors.Mask(semantic_labels.squeeze())
+
+    def _export_semantic_label_masks(self) -> None:
+        """Exports the segmentation binary masks (one-hot) to semantic labels."""
+        total_samples = len(self._samples_dirs)
+        for sample_index in tqdm.trange(
+            total_samples, desc=">> Exporting optimized semantic masks"
+        ):
+            masks_dir = self._get_masks_dir(sample_index)
+            filename = os.path.join(masks_dir, "semantic_labels", "masks.nii.gz")
+            if os.path.isfile(filename):
+                continue
+
+            mask_paths = [os.path.join(masks_dir, label + ".nii.gz") for label in self.classes]
```

Comment:
The same 4 lines of code are used above in `_load_mask `. Would it make sense to refactor those into a function?

Reply:
right I was just doing that :D 

done